### ISSUE TITLE
vm_compute segfaults / Anomaly with cofix

### ISSUE THREAD

**Issue by roconnor** [coq/coq#7333]
#### Version
The Coq Proof Assistant, version 8.8.0 (April 2018)
compiled on Apr 18 2018 15:14:15 with OCaml 4.05.0

and

The Coq Proof Assistant, version 8.7.2 (March 2018)
compiled on Mar 4 2018 20:17:32 with OCaml 4.04.2

#### Operating system

Linux 4.14.29 #1-NixOS SMP x86_64 GNU/Linux

#### Description of the problem

The following (minimized) file causes a segfault when executing the last line:

```coq
CoInductive wrap : Type :=
 | item : unit -> wrap.

Definition extract (t : wrap) : unit :=
match t with
| item x => x
end.

CoFixpoint close u : unit -> wrap :=
match u with
| tt => item
end.

Definition table : wrap := close tt tt.

Eval vm_compute in (extract table).
Eval vm_compute in (extract table). (* Segfault on the second vm_compute *)
```

Alternately the follow file causes an `Anomaly "Uncaught exception Not_found."`:

```coq

Set Primitive Projections.
CoInductive wrap : Type :=
 item { extract : unit }.

CoFixpoint close u : unit -> wrap := 
match u with 
| tt => item
end.

Definition table : wrap := close tt tt.

Eval vm_compute in (extract table).
Eval vm_compute in (extract table). (* Anomaly on the second vm_compute *)
```

**Comment by maximedenes** [comment]
I'm having a look now.

**Comment by maximedenes** [comment]
Ok, this is a serious bug. I already derived a proof of false using `native_compute`, from this example. `vm_compute` is probably going to follow the same path.

**Comment by ppedrot** [comment]
@maximedenes Can you post your proof of false, out of curiosity?

**Comment by maximedenes** [comment]
What are you ready to give? :)

**Comment by Zimmi48** [comment]
The critical bug found by @maximedenes is there since the origins of `native_compute` (8.5). The original reported bug is there even in 8.4 so maybe @maximedenes can cook us a critical bug using `vm_compute` which would work even with 8.4.

**Comment by maximedenes** [comment]
The second example does not follow exactly the same codepath, so there may be two bugs. It is easier to derive a `vm_compute` proof of false from the second (although already a bit involved):

```coq
Set Primitive Projections.

CoInductive wrap : Type :=
 item { extract : bool; extract2: bool; extract3: bool; extract4: bool }.

CoFixpoint close u : bool -> bool -> bool -> bool -> wrap := 
match u with 
| tt => item
end.

Definition table : wrap := close tt false false false true.

Lemma bad : False.
Proof.
assert (true = false).
  change (extract4 table = false).
  vm_compute.
  reflexivity.
discriminate.
Qed.
```

**Comment by maximedenes** [comment]
I don't think this will "work" on old versions, since IIRC the reduction rule was not applying for coinductive records.

**Comment by ppedrot** [comment]
Interesting, I have to evaluate `vm_compute` twice to make the proof go through!

**Comment by maximedenes** [comment]
> Interesting, I have to evaluate vm_compute twice to make the proof go through!

Yep, that's expected :)

**Comment by Zimmi48** [comment]
Yes, this one only "works" starting with 8.6.

**Comment by maximedenes** [comment]
Ok, this one is much trickier but will be interesting to double check once fixed. It required something looking very much like return-oriented programming :)

@Zimmi48 I would be interested in knowing what Coq 8.4 thinks about it (the last minor release of 8.4, because it is interacting with another soundness issue fixed at the time).

```coq
CoInductive t :=
| C_0 : bool -> t
| C_1 : bool -> t
| C_2 : bool -> t
| C_3 : bool -> t
| C_4 : bool -> t
| C_5 : bool -> t
| C_6 : bool -> t
| C_7 : bool -> t
| C_8 : bool -> t
| C_9 : bool -> t
| C_10 : bool -> t
| C_11 : bool -> t
| C_12 : bool -> t
| C_13 : bool -> t
| C_14 : bool -> t
| C_15 : bool -> t
| C_16 : bool -> t
| C_17 : bool -> t
| C_18 : bool -> t
| C_19 : bool -> t
| C_20 : bool -> t
| C_21 : bool -> t
| C_22 : bool -> t
| C_23 : bool -> t
| C_24 : bool -> t
| C_25 : bool -> t
| C_26 : bool -> t
| C_27 : bool -> t
| C_28 : bool -> t
| C_29 : bool -> t
| C_30 : bool -> t
| C_31 : bool -> t
| C_32 : bool -> t
| C_33 : bool -> t
| C_34 : bool -> t
| C_35 : bool -> t
| C_36 : bool -> t
| C_37 : bool -> t
| C_38 : bool -> t
| C_39 : bool -> t
| C_40 : bool -> t
| C_41 : bool -> t
| C_42 : bool -> t
| C_43 : bool -> t
| C_44 : bool -> t
| C_45 : bool -> t
| C_46 : bool -> t
| C_47 : bool -> t
| C_48 : bool -> t
| C_49 : bool -> t
| C_50 : bool -> t
| C_51 : bool -> t
| C_52 : bool -> t
| C_53 : bool -> t
| C_54 : bool -> t
| C_55 : bool -> t
| C_56 : bool -> t
| C_57 : bool -> t
| C_58 : bool -> t
| C_59 : bool -> t
| C_60 : bool -> t
| C_61 : bool -> t
| C_62 : bool -> t
| C_63 : bool -> t
| C_64 : bool -> t
| C_65 : bool -> t
| C_66 : bool -> t
| C_67 : bool -> t
| C_68 : bool -> t
| C_69 : bool -> t
| C_70 : bool -> t
| C_71 : bool -> t
| C_72 : bool -> t
| C_73 : bool -> t
| C_74 : bool -> t
| C_75 : bool -> t
| C_76 : bool -> t
| C_77 : bool -> t
| C_78 : bool -> t
| C_79 : bool -> t
| C_80 : bool -> t
| C_81 : bool -> t
| C_82 : bool -> t
| C_83 : bool -> t
| C_84 : bool -> t
| C_85 : bool -> t
| C_86 : bool -> t
| C_87 : bool -> t
| C_88 : bool -> t
| C_89 : bool -> t
| C_90 : bool -> t
| C_91 : bool -> t
| C_92 : bool -> t
| C_93 : bool -> t
| C_94 : bool -> t
| C_95 : bool -> t
| C_96 : bool -> t
| C_97 : bool -> t
| C_98 : bool -> t
| C_99 : bool -> t
| C_100 : bool -> t
| C_101 : bool -> t
| C_102 : bool -> t
| C_103 : bool -> t
| C_104 : bool -> t
| C_105 : bool -> t
| C_106 : bool -> t
| C_107 : bool -> t
| C_108 : bool -> t
| C_109 : bool -> t
| C_110 : bool -> t
| C_111 : bool -> t
| C_112 : bool -> t
| C_113 : bool -> t
| C_114 : bool -> t
| C_115 : bool -> t
| C_116 : bool -> t
| C_117 : bool -> t
| C_118 : bool -> t
| C_119 : bool -> t
| C_120 : bool -> t
| C_121 : bool -> t
| C_122 : bool -> t
| C_123 : bool -> t
| C_124 : bool -> t
| C_125 : bool -> t
| C_126 : bool -> t
| C_127 : bool -> t
| C_128 : bool -> t
| C_129 : bool -> t
| C_130 : bool -> t
| C_131 : bool -> t
| C_132 : bool -> t
| C_133 : bool -> t
| C_134 : bool -> t
| C_135 : bool -> t
| C_136 : bool -> t
| C_137 : bool -> t
| C_138 : bool -> t
| C_139 : bool -> t
| C_140 : bool -> t
| C_141 : bool -> t
| C_142 : bool -> t
| C_143 : bool -> t
| C_144 : bool -> t
| C_145 : bool -> t
| C_146 : bool -> t
| C_147 : bool -> t
| C_148 : bool -> t
| C_149 : bool -> t
| C_150 : bool -> t
| C_151 : bool -> t
| C_152 : bool -> t
| C_153 : bool -> t
| C_154 : bool -> t
| C_155 : bool -> t
| C_156 : bool -> t
| C_157 : bool -> t
| C_158 : bool -> t
| C_159 : bool -> t
| C_160 : bool -> t
| C_161 : bool -> t
| C_162 : bool -> t
| C_163 : bool -> t
| C_164 : bool -> t
| C_165 : bool -> t
| C_166 : bool -> t
| C_167 : bool -> t
| C_168 : bool -> t
| C_169 : bool -> t
| C_170 : bool -> t
| C_171 : bool -> t
| C_172 : bool -> t
| C_173 : bool -> t
| C_174 : bool -> t
| C_175 : bool -> t
| C_176 : bool -> t
| C_177 : bool -> t
| C_178 : bool -> t
| C_179 : bool -> t
| C_180 : bool -> t
| C_181 : bool -> t
| C_182 : bool -> t
| C_183 : bool -> t
| C_184 : bool -> t
| C_185 : bool -> t
| C_186 : bool -> t
| C_187 : bool -> t
| C_188 : bool -> t
| C_189 : bool -> t
| C_190 : bool -> t
| C_191 : bool -> t
| C_192 : bool -> t
| C_193 : bool -> t
| C_194 : bool -> t
| C_195 : bool -> t
| C_196 : bool -> t
| C_197 : bool -> t
| C_198 : bool -> t
| C_199 : bool -> t
| C_200 : bool -> t
| C_201 : bool -> t
| C_202 : bool -> t
| C_203 : bool -> t
| C_204 : bool -> t
| C_205 : bool -> t
| C_206 : bool -> t
| C_207 : bool -> t
| C_208 : bool -> t
| C_209 : bool -> t
| C_210 : bool -> t
| C_211 : bool -> t
| C_212 : bool -> t
| C_213 : bool -> t
| C_214 : bool -> t
| C_215 : bool -> t
| C_216 : bool -> t
| C_217 : bool -> t
| C_218 : bool -> t
| C_219 : bool -> t
| C_220 : bool -> t
| C_221 : bool -> t
| C_222 : bool -> t
| C_223 : bool -> t
| C_224 : bool -> t
| C_225 : bool -> t
| C_226 : bool -> t
| C_227 : bool -> t
| C_228 : bool -> t
| C_229 : bool -> t
| C_230 : bool -> t
| C_231 : bool -> t
| C_232 : bool -> t
| C_233 : bool -> t
| C_234 : bool -> t
| C_235 : bool -> t
| C_236 : bool -> t
| C_237 : bool -> t
| C_238 : bool -> t
| C_239 : bool -> t
| C_240 : bool -> t
| C_241 : bool -> t
| C_242 : bool -> t
| C_243 : bool -> t
| C_244 : bool -> t
| C_245 : bool -> t
| C_246 : bool -> t
| C_247 : bool -> t
| C_248 : bool -> t
| C_249 : bool -> t
| C_250 : bool -> t
| C_251 : bool -> t
| C_252 : bool -> t
| C_253 : bool -> t
| C_254 : bool -> t
| C_255 : bool -> t
| C_256 : bool -> t.

Definition is_256 x : bool -> bool -> bool :=
  match x with
  | C_1 _ => fun x y => x
  | _ => fun x y => y
  end.

CoFixpoint close u :=
  match u with
    tt => C_1
  end.

Definition table : t := close tt true.

Eval vm_compute in is_256 table true false.

Lemma bad : False.
Proof.
assert (true = false).
  change (is_256 table true false = false).
  vm_compute; reflexivity.
discriminate.
Qed.
```

**Comment by ppedrot** [comment]
@maximedenes It looks like 8.4 happily accepts this proof.

To self: OK, so it's 2018 and buffer overflow is still a thing???

**Comment by ppedrot** [comment]
@maximedenes Also, slightly hijacking the thread, but do we have some statistics on the number of unsoundness bugs w.r.t. the kernel component involved? The VM and native compiler seem to be ranking high on that list...

**Comment by maximedenes** [comment]
> To self: OK, so it's 2018 and buffer overflow is still a thing???

IIUC, it's not really an overflow, on the contrary, I make sure we generate enough code for the unexpected jump (due to `closure_tag`) to land somewhere. Then, I had to fix the stack frames because we execute a `return`. That's why `is_C1` returns functions.

> Also, slightly hijacking the thread, but do we have some statistics on the number of unsoundness bugs w.r.t. the kernel component involved? The VM and native compiler seem to be ranking high on that list...

I don't have statistics. My feeling is that the bytecode and native compilers are probably in the same ballpark as inductive types (with fixpoints). Of course, it would be nice to have a verified compiler.

**Comment by Zimmi48** [comment]
@maximedenes Don't forget the fix for the `v8.8` branch ;-)

**Comment by maximedenes** [comment]
Yes, I plan to do it later this week, but I have a few other things to do first, including going through the pile of PRs waiting for a review from me.
